### PR TITLE
[IOTDB-3445][IOTDB-3446] Fixed empty dirs are not deleted after data …

### PR DESCRIPTION
**Pull request by cmlmakahts**
content:
1. delete empty storage group dirs after all data files are deleted.
2. modify deletion of data region:
     abort compaction
     delete data files
     delete system dir
3. combine _predeletedStorageGroup_ with StorageGroupPartitionTables

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 1]
```diff
@@ -1618,8 +1619,15 @@ public void deleteFolder(String systemDir) {
...
     try {
-      File storageGroupFolder = SystemFileFactory.INSTANCE.getFile(systemDir, dataRegionId);
-      if (storageGroupFolder.exists()) {
+      File dataRegionFilder =
+          SystemFileFactory.INSTANCE.getFile(
+              systemDir + File.separator + logicalStorageGroupName, dataRegionId);
+      if (dataRegionFilder.exists()) {
+        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(dataRegionFilder);
+      }
+      final File storageGroupFolder = dataRegionFilder.getParentFile();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1628]:
```suggestion
      final File storageGroupSchemaFolder = dataRegionFilder.getParentFile();
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 2]
```diff
@@ -1618,8 +1619,15 @@ public void deleteFolder(String systemDir) {
         systemDir);
     writeLock("deleteFolder");
     try {
-      File storageGroupFolder = SystemFileFactory.INSTANCE.getFile(systemDir, dataRegionId);
-      if (storageGroupFolder.exists()) {
+      File dataRegionFilder =
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1622]:
```suggestion
      File dataRegionSchemaFolder =
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 3]
```diff
@@ -1674,9 +1682,14 @@ public void syncDeleteDataFiles() {
 
   private void deleteAllSGFolders(List<String> folder) {
     for (String tsfilePath : folder) {
-      File storageGroupFolder =
+      File dataRegionFolder =
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1685]:
```suggestion
      File dataRegionDataFolder =
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 4]
```diff
@@ -1674,9 +1682,14 @@ public void syncDeleteDataFiles() {
...
           fsFactory.getFile(tsfilePath, logicalStorageGroupName + File.separator + dataRegionId);
-      if (storageGroupFolder.exists()) {
+      if (dataRegionFolder.exists()) {
+        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(dataRegionFolder);
+      }
+      final File storageGroupFolder = dataRegionFolder.getParentFile();
+      if (storageGroupFolder.isDirectory()
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1691]:
```suggestion
      if (storageGroupDataFolder.isDirectory()
```

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 5]
```diff
@@ -302,6 +295,11 @@ public DataSet getDataPartition(GetDataPartitionReq req) {
     return dataPartitionResp;
   }
 
+  private boolean isStorageGroupExisted(String storageGroup) {
+    return storageGroupPartitionTables.containsKey(storageGroup)
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:299]: better to get one time

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 6]
```diff
@@ -1618,9 +1619,16 @@ public void deleteFolder(String systemDir) {
...
     writeLock("deleteFolder");
     try {
-      File storageGroupFolder = SystemFileFactory.INSTANCE.getFile(systemDir, dataRegionId);
-      if (storageGroupFolder.exists()) {
-        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(storageGroupFolder);
+      File dataRegionSystemFolder =
+          SystemFileFactory.INSTANCE.getFile(
+              systemDir + File.separator + logicalStorageGroupName, dataRegionId);
+      if (dataRegionSystemFolder.exists()) {
+        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(dataRegionSystemFolder);
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1626]: package name?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 7]
```diff
@@ -1618,9 +1619,16 @@ public void deleteFolder(String systemDir) {
...
+      }
+      final File storageGroupSystemFolder = dataRegionSystemFolder.getParentFile();
+      if (storageGroupSystemFolder.isDirectory()
+          && Objects.requireNonNull(storageGroupSystemFolder.listFiles()).length == 0) {
+        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(storageGroupSystemFolder);
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1631]: the same as above

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java (hunk 6 above)
Comment (cmlmakahts) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1626]: Here are two FileUtils introduced, 

image

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 8]
```diff
@@ -350,18 +347,20 @@ public DataSet getSchemaNodeManagementPartition(List<String> matchedStorageGroup
...
+              storageGroupPartitionTables
+                  .get(storageGroup)
+                  .getSchemaPartition(new ArrayList<>(), schemaPartitionMap.get(storageGroup));
 
-          if (schemaPartitionMap.get(storageGroup).size() == 0) {
-            // Remove empty Map
-            schemaPartitionMap.remove(storageGroup);
-          }
-        });
+              if (schemaPartitionMap.get(storageGroup).size() == 0) {
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:359]: space？

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 9]
```diff
@@ -45,7 +45,7 @@
 import java.util.concurrent.atomic.AtomicInteger;
 
 public class StorageGroupPartitionTable {
-
+  private boolean isPredeleted = false;
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:48]: need add volatile?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 8 above)
Comment (cmlmakahts) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:359]: Code align is unified by _spotless_.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 9 above)
Comment (cmlmakahts) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:48]: Fixed.

**Review by CRZbulabula — APPROVED** [review]
LGTM~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 5 above)
Comment (cmlmakahts) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java (hunk 4 above)
Comment (cmlmakahts) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java (hunk 3 above)
Comment (cmlmakahts) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java (hunk 1 above)
Comment (cmlmakahts) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 10]
```diff
@@ -1674,10 +1682,15 @@ public void syncDeleteDataFiles() {
 
   private void deleteAllSGFolders(List<String> folder) {
     for (String tsfilePath : folder) {
-      File storageGroupFolder =
+      File dataRegionDataFolder =
           fsFactory.getFile(tsfilePath, logicalStorageGroupName + File.separator + dataRegionId);
-      if (storageGroupFolder.exists()) {
-        org.apache.iotdb.commons.utils.FileUtils.deleteDirectory(storageGroupFolder);
+      if (dataRegionDataFolder.exists()) {
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:1687]: Wrap these codes as a function?

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionUtils.java [hunk 11]
```diff
@@ -61,6 +62,10 @@ public static void deleteSchemaRegionFolder(String schemaRegionDirPath, Logger l
           String.format(
               "Failed to delete schema region folder %s", schemaRegionDir.getAbsolutePath()));
     }
+    final File storageGroupDir = schemaRegionDir.getParentFile();
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionUtils.java:65]: same

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionUtils.java (hunk 11 above)
Comment (cmlmakahts) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionUtils.java:65]: Situation here is a bit different.  A MetadataException is thrown when deletion fail.  I suggest keep the way it is here.

File: node-commons/src/main/java/org/apache/iotdb/commons/utils/FileUtils.java [hunk 12]
```diff
@@ -56,6 +57,15 @@ public static void deleteDirectory(File folder) {
     }
   }
 
+  public static void deleteDirectoryAndParent(File folder) {
```
Comment (qiaojialin) [inline node-commons/src/main/java/org/apache/iotdb/commons/utils/FileUtils.java:60]:
```suggestion
  public static void deleteDirectoryAndEmptyParent(File folder) {
```

**Pull request merged**